include/roboteam_ai/stp/Role.hpp

```diff
@@ -15,57 +15,69 @@ namespace rtt::ai::stp {
  * Role class used in STP, pretty self explanatory
  * essentially a state machine of Tactics
  */
-    class Role {
-    public:
-        Role(std::string name)
-                : roleName{std::move(name)} {}
-
-        /**
-         * Function that's called every tick, default implementation is robotTactics.update();
-         * @param info TacticInfo to be passed to update()
-         * @return The status that the current tactic returns
-         */
-        [[nodiscard]] virtual Status update(StpInfo const &info) noexcept;
-
-        /**
-         * @return True if all tactics returned Status::finish
-         */
-        [[nodiscard]] bool finished() const noexcept;
-
-        /**
-         * Gets the name
-         * @return name of the role
-         */
-        std::string getName() { return roleName; }
-
-        /**
-         * Gets the current robot
-         * @return view to the robot this role belongs to, optional.
-         */
-         [[nodiscard]] std::optional<world_new::view::RobotView> const& getCurrentRobot() const;
-
-         /**
-          * Gets the tactic whose turn it is
-          * @return Tactic*
-          */
-          [[nodiscard]] Tactic* getCurrentTactic();
-
-    protected:
-        /**
-         * Robot to which this role is currently assigned
-         */
-        std::optional<world_new::view::RobotView> currentRobot;
-
-        /**
-         * Name of the role
-         */
-        std::string roleName{};
-
-        /**
-         * State machine that keeps track of tactic states
-         */
-        collections::state_machine<Tactic, Status, StpInfo> robotTactics;
-    };
+class Role {
+public:
+    Role(std::string name)
+        : roleName{std::move(name)} {}
+
+    /**
+    * Function that's called every tick, default implementation is robotTactics.update();
+    * @param info TacticInfo to be passed to update()
+    * @return The status that the current tactic returns
+    */
+    [[nodiscard]] virtual Status update(StpInfo const &info) noexcept;
+
+    /**
+    * @return True if all tactics returned Status::finish
+    */
+    [[nodiscard]] bool finished() const noexcept;
+
+    /**
+    * Gets the name
+    * @return name of the role
+    */
+    std::string getName() { return roleName; }
+
+    /**
+    * Gets the current robot
+    * @return view to the robot this role belongs to, optional.
+    */
+    [[nodiscard]] std::optional<world_new::view::RobotView> const& getCurrentRobot() const;
+
+    /**
+    * Gets the tactic whose turn it is
+    * @return Tactic*
+    */
+    [[nodiscard]] Tactic* getCurrentTactic();
+
+    /**
+    * Forces the Role to skip to the next tactic in the state machine
+    */
+    void forceNextTactic() noexcept;
+
+    /**
+     * Forces the Role to skip to tactic n in the state machine
+     * @param n Index of the tactic in the state machine
+     */
+    void forceToTactic(int n) noexcept;
```

Comment:
Is this really needed? Usually you either reset or force next tactic, which is more readable

Reply:
The reason I did this is that the defender role has two tactics, blockrobot and intercept. For the play we want all the defenders to block a robot, except that whenever the ball is moving, the closest defender will try to intercept the ball, for example. In that case the play can tell that specific robot to skip to the intercept tactic. The problem with forcing next tactic is then that the play has to know which tactic the robot is executing, to avoid constantly forcing to the next tactic.

Reply:
In this particular case you do know, as the Role will no longer return Waiting after forcing the next tactic